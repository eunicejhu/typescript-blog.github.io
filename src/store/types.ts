import { PostsState } from "../features/posts/postsSlice";
import { NotificationsState } from "../features/notifications/notificationsSlice";
import { UsersState } from "../features/users/usersSlice";
export type State = {
  posts: PostsState;
  users: UsersState;
  notifications: NotificationsState;
};
