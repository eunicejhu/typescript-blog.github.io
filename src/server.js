// eslint-disable-next-line import/no-extraneous-dependencies
import { createServer, Factory, Model } from "miragejs";
import { sub } from "date-fns";

export const makeServer = ({ environment = "test" } = {}) => {
  let server = createServer({
    environment,
    models: { user: Model, post: Model, notification: Model },
    factories: {
      user: Factory.extend({
        name(i) {
          return ["Tianna Jenkins", "Kevin Grant", "Madison Price"][i % 3];
        },
      }),
      // id starts from "1", i starts with 0
      post: Factory.extend({
        title(i) {
          const candidates = ["First", "Second"];
          return `${candidates[i % 2]} test Post`;
        },
        content: "test",
        userId() {
          const candidates = [1, 2, 3];
          return `${candidates[Math.floor(Math.random() * 3)]}`;
        },
        date(i) {
          return sub(new Date(), { days: i }).toISOString();
        },
        reactions(i) {
          const candidates = [
            { thumbsUp: 0, hooray: 0, heart: 4, rocket: 0, eyes: 0 },
            { thumbsUp: 3, hooray: 0, heart: 0, rocket: 0, eyes: 0 },
          ];
          return candidates[i % 2];
        },
      }),
      notification: Factory.extend({
        date(i) {
          // eslint-disable-next-line no-param-reassign
          return sub(new Date(), { days: i }).toISOString();
        },
        userId(i) {
          const candidates = [1, 2, 3];
          return `${candidates[Math.floor(Math.random() * 3)]}`;
        },
        message(i) {
          const candidates = [
            "Glad to know you",
            "Concert in two weeks",
            "Cinema next thursday",
          ];
          return candidates[i % 3];
        },
      }),
    },
    seeds(server) {
      server.createList("user", 3);
      server.createList("notification", 30);
      server.createList("post", 20);
    },
    routes() {
      this.namespace = "api";
      this.get("/users", (schema) => {
        return schema.users.all();
      });
      this.get("/posts", (schema) => {
        return schema.posts.all();
      });
      this.get("/notifications", (schema) => {
        return schema.notifications.all();
      });

      this.post("/posts", (schema, request) => {
        const { requestBody } = request;
        const { data } = JSON.parse(requestBody);
        const { id: postId, reactions } = data;

        if (postId) {
          // add Reaction
          const post = schema.posts.find(postId);
          post.update("reactions", reactions);
          return post;
        } else {
          //create Post
          const date = new Date().toISOString();
          return schema.posts.create({
            ...data,
            date,
            reactions: { thumbsUp: 0, hooray: 0, heart: 0, rocket: 0, eyes: 0 },
          });
        }
      });

      this.put("/posts", (schema, request) => {
        const { requestBody } = request;
        const {
          data: { id, ...attrs },
        } = JSON.parse(requestBody);

        const existingPost = schema.posts.find(id);
        Object.entries(attrs).forEach(([attr, value]) => {
          existingPost.update(attr, value);
        });
        return existingPost;
      });
    },
  });
  return server;
};
