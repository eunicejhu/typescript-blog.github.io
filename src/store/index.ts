import { configureStore } from "@reduxjs/toolkit";
import postsReducer from "../features/posts/postsSlice";
import usersReducer from "../features/users/usersSlice";
import notificationsReducer from "../features/notifications/notificationsSlice";
import { useDispatch } from "react-redux";

const store = configureStore({
  reducer: {
    posts: postsReducer,
    users: usersReducer,
    notifications: notificationsReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();

export type State = ReturnType<typeof store.getState>;

export default store;
