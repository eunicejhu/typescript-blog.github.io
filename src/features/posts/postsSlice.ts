import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import Client from "../../api/client";
import { State } from "../../store/index";
const UPDATE_POST_ERROR_MSG = "Failed to update post";
export const FETCH_POSTS_ERROR_MSG = "Failed to fetch posts";
export type PostsState = {
  data: Post[];
  status: "idle" | "loading" | "succeeded" | "failed";
  error: string | undefined;
  updatePostError: string | undefined;
};
export interface Post {
  id: string;
  title: string;
  content: string;
  userId: string;
  date: string;
  reactions: {
    thumbsUp: number;
    hooray: number;
    heart: number;
    rocket: number;
    eyes: number;
  };
}

export interface AddNewPostAction {
  type: string;
  payload: Post;
}

export interface PostUpdatedAction {
  type: string;
  payload: Post;
}

export interface ReactionAddedAction {
  type: string;
  payload: {
    postId: string;
    reaction: keyof Post["reactions"];
  };
}

export const fetchPosts = createAsyncThunk("posts/fetchPosts", async () => {
  let res;
  res = await Client.fetchPost<Post[]>();
  return res;
});

type addNewPostThunkArg = { title: string; content: string; userId: string };
export const addNewPost = createAsyncThunk(
  "posts/addNewPost",
  async ({ title, content, userId }: addNewPostThunkArg) => {
    let res;
    res = await Client.addNewPost<Post>({
      title,
      content,
      userId,
    });
    return res;
  }
);

type updatePostThunkArg = { id: string; title: string; content: string };
export const updatePost = createAsyncThunk(
  "posts/updatePost",
  async ({ id, title, content }: updatePostThunkArg) => {
    let res;
    res = await Client.updatePost<Post>({
      id,
      title,
      content,
    });
    return res;
  }
);

export type addReactionThunkArg = {
  postId: string;
  reaction: keyof Post["reactions"];
};
export const addReaction = createAsyncThunk(
  "posts/addReaction",
  async ({ postId, reaction }: addReactionThunkArg, { getState }) => {
    let res;
    const reactions = (getState() as State).posts.data.find(
      (post) => post.id === postId
    )?.reactions;
    if (reactions) {
      const updatedReactions: Post["reactions"] = {
        ...reactions,
        [reaction]: reactions[reaction] + 1,
      };
      res = await Client.addReaction<Post>({
        id: postId,
        reactions: updatedReactions,
      });
      return res;
    } else {
      throw Error(
        `posts/addReaction: post (id: ${postId}) does not exist in store. BEFORE Rest API call `
      );
    }
  }
);
const initialState: PostsState = {
  data: [],
  status: "idle",
  error: undefined,
  updatePostError: undefined,
};
const postsSlice = createSlice({
  name: "posts",
  initialState: initialState,
  reducers: {},
  extraReducers: (builder) => {
    // fetchPost
    builder.addCase(fetchPosts.pending, (state) => {
      state.status = "loading";
    });
    builder.addCase(fetchPosts.fulfilled, (state, action) => {
      state.data = action.payload.data;
      state.status = "succeeded";
      state.error = undefined;
    });
    builder.addCase(fetchPosts.rejected, (state, action) => {
      let errorMessageForDev = `${action.type}: Error from REST API (${action.error.message})`;
      console.error(errorMessageForDev);
      state.error = FETCH_POSTS_ERROR_MSG;
      state.status = "failed";
      state.data = [];
    });
    // addNewPost
    builder.addCase(addNewPost.fulfilled, (state, action) => {
      state.data.push(action.payload.data);
    });
    // updatePost
    builder.addCase(updatePost.fulfilled, (state, action) => {
      let { id, ...rest } = action.payload.data;
      const existingPost = state.data.find((post) => post.id === id);
      if (existingPost) {
        Object.entries(rest).forEach(([attr, value]) => {
          (existingPost as Post)[attr as keyof Post] = value;
        });
      } else {
        // TODO: log detailed error message for developer
        let errorMessageForDev = `${
          action.type
        }: Error from client: existingPost not found for data from server ${JSON.stringify(
          action.payload
        )}`;
        console.error(errorMessageForDev);

        let errorMessageForUser = UPDATE_POST_ERROR_MSG;
        state.updatePostError = errorMessageForUser;
      }
    });
    builder.addCase(updatePost.rejected, (state, action) => {
      let errorMessageForDev = `${action.type}: Error from REST API ${action.error.message}`;
      console.error(errorMessageForDev);
      let errorMessageForUser = UPDATE_POST_ERROR_MSG;
      state.updatePostError = errorMessageForUser;
    });
    // addReaction
    builder.addCase(addReaction.fulfilled, (state, action) => {
      const { id, ...rest } = action.payload.data;
      const existingPost = state.data.find((post) => post.id === id);
      if (existingPost) {
        Object.entries(rest).forEach(([attr, value]) => {
          (existingPost as Post)[attr as keyof Post] = value;
        });
      } else {
        let errorMessageForDev = `${
          action.type
        }: Error from client: existingPost not found for data from server ${JSON.stringify(
          action.payload.data
        )}. AFTER REST api call`;
        console.error(errorMessageForDev);
      }
    });
    builder.addCase(addReaction.rejected, (state, action) => {
      let errorMessageForDev = `${action.type}: Error from REST API ${action.error.message}`;
      console.error(errorMessageForDev);
    });
  },
});

const { reducer } = postsSlice;

export default reducer;
