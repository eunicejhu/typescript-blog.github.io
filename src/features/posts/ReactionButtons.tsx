import React from "react";
import { useDispatch } from "react-redux";
import { Post, addReaction } from "./postsSlice";

const reactionEmoji = {
  thumbsUp: "👍",
  hooray: "🎉",
  heart: "❤️",
  rocket: "🚀",
  eyes: "👀",
};

type Props = { post: Post };
type Name = keyof Props["post"]["reactions"];
const ReactionButtons: React.FC<Props> = ({
  post: { reactions, id },
}: Props) => {
  const dispatch = useDispatch();
  const reactionButtons = Object.entries(reactionEmoji).map(([name, emoji]) => (
    <button
      key={name}
      type="button"
      name={name}
      data-testid={name}
      onClick={() =>
        dispatch(addReaction({ postId: id, reaction: name as Name }))
      }
    >
      {emoji} {reactions[name as Name]}
    </button>
  ));
  return <div className="reactions-container">{reactionButtons}</div>;
};

export default ReactionButtons;
