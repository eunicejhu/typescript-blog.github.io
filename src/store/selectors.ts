// selectors
import { State } from "./index";
import { Notification } from "../features/notifications/notificationsSlice";
import { Post } from "../features/posts/postsSlice";
import { User } from "../features/users/usersSlice";

export const selectAllPosts = (state: State): Post[] => state.posts.data;
export const selectPostById = (
  state: State,
  postId: string
): Post | undefined => state.posts.data.find((post) => post.id === postId);
export const selectAllNotifications = (state: State): Notification[] =>
  state.notifications;
export const selectAllUsers = (state: State): User[] => state.users.data;
