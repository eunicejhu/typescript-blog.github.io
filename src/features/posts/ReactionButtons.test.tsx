import React from "react";
import { fireEvent, render } from "@testing-library/react";
// eslint-disable-next-line testing-library/no-dom-import
import { waitFor } from "@testing-library/dom";
import { Provider } from "react-redux";
import ReactionButtons from "./ReactionButtons";
import { fetchPosts } from "./postsSlice";
import store from "../../store";

import { makeServer } from "../../server";
import { Server } from "miragejs";
let server: Server;
describe("ReactionButtons test", () => {
  beforeEach(() => {
    server = makeServer();
    server.createList("post", 2);
    store.dispatch(fetchPosts());
  });
  afterEach(() => {
    server.shutdown();
  });
  it("render correctly", async () => {
    await waitFor(() => {
      expect(store.getState().posts.data.length).toBe(2);
    });
    const post = store.getState().posts.data[0];
    const ui = (
      <Provider store={store}>
        <ReactionButtons post={post} />
      </Provider>
    );
    const { asFragment } = render(ui);
    expect(asFragment()).toMatchSnapshot();
  });

  it("click reaction", async () => {
    await waitFor(() => {
      expect(store.getState().posts.data.length).toBe(2);
    });
    let post;
    post = store.getState().posts.data[0];
    const ui = (
      <Provider store={store}>
        <ReactionButtons post={post} />
      </Provider>
    );
    const { getByTestId, rerender } = render(ui);
    const heartButton = getByTestId("heart");

    fireEvent.click(heartButton);
    await waitFor(() => {
      expect(store.getState().posts.data[0].reactions["heart"]).toBe(5);
    });

    // get updated post from store
    post = store.getState().posts.data[0];
    rerender(
      <Provider store={store}>
        <ReactionButtons post={post} />
      </Provider>
    );
    await waitFor(() => {
      expect(heartButton).toHaveTextContent(/5/i);
    });
  });
});
