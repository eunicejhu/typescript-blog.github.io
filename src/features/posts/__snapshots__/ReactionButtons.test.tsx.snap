// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactionButtons test render correctly 1`] = `
<DocumentFragment>
  <div
    class="reactions-container"
  >
    <button
      data-testid="thumbsUp"
      name="thumbsUp"
      type="button"
    >
      👍 0
    </button>
    <button
      data-testid="hooray"
      name="hooray"
      type="button"
    >
      🎉 0
    </button>
    <button
      data-testid="heart"
      name="heart"
      type="button"
    >
      ❤️ 4
    </button>
    <button
      data-testid="rocket"
      name="rocket"
      type="button"
    >
      🚀 0
    </button>
    <button
      data-testid="eyes"
      name="eyes"
      type="button"
    >
      👀 0
    </button>
  </div>
</DocumentFragment>
`;
